ssl/s3_msg.c

```diff
@@ -52,6 +52,8 @@ int ssl3_send_alert(SSL *s, int level, int desc)
                                           * protocol_version alerts */
     if (desc < 0)
         return -1;
+    if (s->shutdown & SSL_SENT_SHUTDOWN && desc != SSL_AD_CLOSE_NOTIFY)
+        return -1;
     /* If a fatal one, remove from cache */
     if ((level == SSL3_AL_FATAL) && (s->session != NULL))
         SSL_CTX_remove_session(s->session_ctx, s->session);
```

Comment:
There is perhaps a subtle philosophical question of whether attempting to send a fatal alert after shutdown should remove the session from the cache ... I am not sure there is a clear right answer, so the answer that's here is probably tolerable.

Reply:
I also hesitated here

Reply:
Given that the our side of the connection can get closed while the data was in transit, I don't think it should.
Consider an HTTP server processing a keep-alive session, with a HTTP-level timeout triggering because client didn't send a new request while the new request was in flight. In such situation getting App Data after sending `close_notify` is expected, and not a result of misbehaviour or an attack.